lib/Parse/ParseType.cpp

```diff
@@ -722,9 +723,16 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
       SourceLoc LAngle, RAngle;
       SmallVector<TypeRepr*, 8> GenericArgs;
       if (startsWithLess(Tok)) {
-        auto genericArgsStatus = parseGenericArguments(GenericArgs, LAngle, RAngle);
-        if (genericArgsStatus.isErrorOrHasCompletion())
-          return genericArgsStatus;
+        // Only attempt to parse a generic argument list in a cast destination
+        // type if the token text is just "<", because it can be an operator,
+        // for example: "1 as Int16 << 7".
+        if (Tok.getText().equals("<") ||
+            reason != ParseTypeReason::CastDestination) {
```

Comment:
Can't we just use `canParseAsGenericArgumentList()` instead of `startsWithLess(Tok)` [like expression parsing](https://github.com/apple/swift/blob/0464fd302fd0e690aab492085530bebea454fcec/lib/Parse/ParseExpr.cpp#L2388)?
It only accepts single `<`, but I'm not sure there's a place we need to accept `<<` for a generic arguments in the first place. If we cannot use `canParseAsGenericArgumentList()`, could you add some code-comment to explain why not?

Reply:
I see we don't use `canParseAsGenericArgumentList()` probably because of the SIL type parsing which has generic function types like `<T> (T) -> U`.  And for non-expression position, we probably want to diagnose errors in `<...>` as a generic argument error, but not as a bogus random tokens after a type. E.g.

```
let a: Dictionary<String: Int>
```
The error in `<String: Int>` should be diagnosed as like `expected ',' instead of ':'` inside `parseGenericArguments()`. But if we use `canParseAsGenericArgumentList()` here, it's parsed as a random tokens after `let a: Dictionary`.

So I think having `ParseTypeReason::CastDestination` actually makes sense. And we _can_ use `canParseAsGenericArgumentList()` only for `CastDestination` case.

Reply:
Thank you for the feedback!